lib/log_dna_batcher.rb

```diff
@@ -0,0 +1,116 @@
+# frozen_string_literal: true
+
+require "net/http"
+require "uri"
+require "json"
+require "socket"
+
+class LogDnaBatcher
+  def initialize(api_key, base_url: "https://logs.logdna.com/logs/ingest",
+    default_metadata: {"app" => $0, "pid" => Process.pid},
+    flush_interval: 0.5,
+    max_batch_size: 5)
+    @api_key = api_key
+    @flush_interval = flush_interval
+    @max_batch_size = max_batch_size
+    @input_queue = Queue.new
+    @http = nil
+    @base_url = URI.parse(base_url)
+    @default_metadata = default_metadata.dup.freeze
+
+    params = @base_url.query ? URI.decode_www_form(@base_url.query).to_h : {}
+    params["hostname"] = Socket.gethostname
+    params["app"] = $0
+    @request_path = "#{@base_url.path}?#{URI.encode_www_form(params)}"
+
+    start_processor
+  end
+
+  def log(line, timestamp: (Time.now.to_f * 1000).to_i, **metadata)
+    data = @default_metadata.dup
+    data.merge!(metadata)
+    data[:line] = line
+    data[:timestamp] = timestamp
+    @input_queue.push(data)
+    nil
+  end
+
+  def stop
+    @input_queue.close
+    @processor_thread&.join
+    close_connection
+  end
+
+  def send_batch(batch)
+    return if batch.empty?
+
+    begin
+      http = ensure_connection
+
+      request = Net::HTTP::Post.new(@request_path)
+      request.basic_auth(@api_key, "")
+      request["Content-Type"] = "application/json; charset=UTF-8"
+      request["Connection"] = "keep-alive"
+      request.body = {lines: batch}.to_json
+
+      response = http.request(request)
+
+      if response.is_a?(Net::HTTPSuccess)
+        batch.clear
+      else
+        puts "Failed to send logs: #{response.code} #{response.message}"
```

Comment:
We should track the number of failures, and after a number of consecutive failures, we should raise a different error that will result in the process exiting.  Otherwise, we run the risk of continuing to run while things are not getting logged.